File: android/src/main/java/com/easemob/im_flutter_sdk/EMContactManagerWrapper.java
package com.easemob.im_flutter_sdk;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.hyphenate.EMContactListener;
import com.hyphenate.chat.EMClient;
import com.hyphenate.exceptions.HyphenateException;

import org.json.JSONException;
import org.json.JSONObject;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;


public class EMContactManagerWrapper extends EMWrapper implements MethodCallHandler{

    EMContactManagerWrapper(PluginRegistry.Registrar registrar, String channelName) {
        super(registrar, channelName);
        registerEaseListener();
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {

        JSONObject param = (JSONObject)call.arguments;
        try {
            if (EMSDKMethod.addContact.equals(call.method))
            {
                addContact(param, EMSDKMethod.addContact, result);
            }
            else if(EMSDKMethod.deleteContact.equals(call.method))
            {
                deleteContact(param, EMSDKMethod.deleteContact, result);
            }
            else if(EMSDKMethod.getAllContactsFromServer.equals(call.method))
            {
                getAllContactsFromServer(param, EMSDKMethod.getAllContactsFromServer, result);
            }
            else if(EMSDKMethod.addUserToBlackList.equals(call.method))
            {
                addUserToBlackList(param, EMSDKMethod.getAllContactsFromServer, result);
            }
            else if(EMSDKMethod.removeUserFromBlackList.equals(call.method))
            {
                removeUserFromBlackList(param, EMSDKMethod.getAllContactsFromServer, result);
            }
            else if(EMSDKMethod.getBlackListFromServer.equals(call.method))
            {
                getBlackListFromServer(param, EMSDKMethod.getAllContactsFromServer, result);
            }
            else if(EMSDKMethod.acceptInvitation.equals(call.method))
            {
                acceptInvitation(param, EMSDKMethod.getAllContactsFromServer, result);
            }
            else if(EMSDKMethod.declineInvitation.equals(call.method))
            {
                declineInvitation(param, EMSDKMethod.getAllContactsFromServer, result);
            }
            else if(EMSDKMethod.getSelfIdsOnOtherPlatform.equals(call.method))
            {
                getSelfIdsOnOtherPlatform(param, EMSDKMethod.getAllContactsFromServer, result);
            }
            else  {
                super.onMethodCall(call, result);
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private void addContact(JSONObject param, String channelName, Result result) throws JSONException {
        String username = param.getString("username");
        String reason = param.getString("reason");

        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().addContact(username, reason);
                onSuccess(result, channelName, username);
            }catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void deleteContact(JSONObject param, String channelName, Result result) throws JSONException {
        String username = param.getString("username");
        boolean keepConversation = param.getBoolean("keepConversation");
        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().deleteContact(username, keepConversation);
                onSuccess(result, channelName, username);
            }catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void getAllContactsFromServer(JSONObject params, String channelName,  Result result) throws JSONException {
        asyncRunnable(() -> {
            try {
                List contacts =  EMClient.getInstance().contactManager().getAllContactsFromServer();
                onSuccess(result, channelName, contacts);
            }catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void addUserToBlackList(JSONObject params, String channelName,  Result result) throws JSONException {
        String username = params.getString("username");
        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().addUserToBlackList(username, false);
                onSuccess(result, channelName, username);
            }catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void removeUserFromBlackList(JSONObject params, String channelName,  Result result) throws JSONException {
        String username = params.getString("username");
        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().removeUserFromBlackList(username);
                onSuccess(result, channelName, username);
            }catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void getBlackListFromServer(JSONObject params, String channelName,  Result result) throws JSONException {
        asyncRunnable(() -> {
            try {
                List contacts = EMClient.getInstance().contactManager().getBlackListFromServer();
                onSuccess(result, channelName, contacts);
            }catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void acceptInvitation(JSONObject params, String channelName,  Result result) throws JSONException {
        String username = params.getString("username");
        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().acceptInvitation(username);
                onSuccess(result, channelName, username);
            }catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void declineInvitation(JSONObject params, String channelName,  Result result) throws JSONException {
        String username = params.getString("username");
        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().declineInvitation(username);
                onSuccess(result, channelName, username);
            }catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void getSelfIdsOnOtherPlatform(JSONObject params, String channelName,  Result result) throws JSONException {
        asyncRunnable(() -> {
            try {
                List platforms = EMClient.getInstance().contactManager().getSelfIdsOnOtherPlatform();
                onSuccess(result, channelName, platforms);
            }catch (HyphenateException e) {
                onError(result, e);
            }
        });

    }

    private void registerEaseListener(){
        EMClient.getInstance().contactManager().setContactListener(new EMContactListener() {
            @Override
            public void onContactAdded(String userName) {
                Map<String, Object> data = new HashMap<>();
                data.put("type", "onContactAdded");
                data.put("username", userName);
                post(() -> channel.invokeMethod(EMSDKMethod.onContactChanged, data));

            }

            @Override
            public void onContactDeleted(String userName) {
                Map<String, Object> data = new HashMap<>();
                data.put("type", "onContactDeleted");
                data.put("username", userName);
                post(() -> channel.invokeMethod(EMSDKMethod.onContactChanged, data));
            }

            @Override
            public void onContactInvited(String userName, String reason) {
                Map<String, Object> data = new HashMap<>();
                data.put("type", "onContactInvited");
                data.put("username", userName);
                data.put("reason", reason);
                post(() -> channel.invokeMethod(EMSDKMethod.onContactChanged, data));
            }

            @Override
            public void onFriendRequestAccepted(String userName) {
                Map<String, Object> data = new HashMap<>();
                data.put("type", "onFriendRequestAccepted");
                data.put("username", userName);
                post(() -> channel.invokeMethod(EMSDKMethod.onContactChanged, data));
            }

            @Override
            public void onFriendRequestDeclined(String userName) {
                Map<String, Object> data = new HashMap<>();
                data.put("type", "onFriendRequestDeclined");
                data.put("username", userName);
                post(() -> channel.invokeMethod(EMSDKMethod.onContactChanged, data));
            }
        });
    }
}

File: ios/Classes/EMContactManagerWrapper.m
//
//  EMContactManagerWrapper.m
//  
//
//  Created by 杜洁鹏 on 2019/10/8.
//

#import "EMContactManagerWrapper.h"
#import "EMSDKMethod.h"

@interface EMContactManagerWrapper () <EMContactManagerDelegate>

@end

@implementation EMContactManagerWrapper
- (instancetype)initWithChannelName:(NSString *)aChannelName
                          registrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    if(self = [super initWithChannelName:aChannelName
                           registrar:registrar]) {
       [EMClient.sharedClient.contactManager addDelegate:self delegateQueue:nil];
    }
    return self;
}


#pragma mark - FlutterPlugin

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    
    if ([EMMethodKeyAddContact isEqualToString:call.method]) {
        [self addContact:call.arguments
             channelName:call.method
                  result:result];
    } else if ([EMMethodKeyDeleteContact isEqualToString:call.method]) {
        [self deleteContact:call.arguments
                channelName:call.method
                     result:result];
    } else if ([EMMethodKeyGetAllContactsFromServer isEqualToString:call.method]) {
        [self getAllContactsFromServer:call.arguments
                           channelName:call.method
                                result:result];
    } else if ([EMMethodKeyGetAllContactsFromDB isEqualToString:call.method]) {
        [self getAllContactsFromDB:call.arguments
                       channelName:call.method
                            result:result];
    } else if ([EMMethodKeyAddUserToBlockList isEqualToString:call.method]) {
        [self addUserToBlockList:call.arguments
                     channelName:call.method
                          result:result];
    } else if ([EMMethodKeyRemoveUserFromBlockList isEqualToString:call.method]) {
        [self removeUserFromBlockList:call.arguments
                          channelName:call.method
                               result:result];
    } else if ([EMMethodKeyGetBlockListFromServer isEqualToString:call.method]) {
        [self getBlockListFromServer:call.arguments
                         channelName:call.method
                              result:result];
    } else if ([EMMethodKeyGetBlockListFromDB isEqualToString:call.method]){
        [self getBlockListFromDB:call.arguments
                     channelName:call.method
                          result:result];
    } else if ([EMMethodKeyAcceptInvitation isEqualToString:call.method]) {
        [self acceptInvitation:call.arguments
                   channelName:call.method
                        result:result];
    } else if ([EMMethodKeyDeclineInvitation isEqualToString:call.method]) {
        [self declineInvitation:call.arguments
                    channelName:call.method
                         result:result];
    } else if ([EMMethodKeyGetSelfIdsOnOtherPlatform isEqualToString:call.method]) {
        [self getSelfIdsOnOtherPlatform:call.arguments
                            channelName:call.method
                                 result:result];
    } else {
        [super handleMethodCall:call result:result];
    }
}


#pragma mark - Actions
- (void)addContact:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    NSString *reason = param[@"reason"];
    [EMClient.sharedClient.contactManager addContact:username
                                             message:reason
                                          completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
}

- (void)deleteContact:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    BOOL keepConversation = [param[@"keepConversation"] boolValue];
    [EMClient.sharedClient.contactManager deleteContact:username
                                   isDeleteConversation:keepConversation
                                             completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
}

- (void)getAllContactsFromServer:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    [EMClient.sharedClient.contactManager getContactsFromServerWithCompletion:^(NSArray *aList, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aList];
    }];
}

- (void)getAllContactsFromDB:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSArray *aList = EMClient.sharedClient.contactManager.getContacts;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:aList];
}


- (void)addUserToBlockList:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    [EMClient.sharedClient.contactManager addUserToBlackList:username
                                                  completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
}

- (void)removeUserFromBlockList:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    [EMClient.sharedClient.contactManager removeUserFromBlackList:username
                                                       completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
    
}

- (void)getBlockListFromServer:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    [EMClient.sharedClient.contactManager getBlackListFromServerWithCompletion:^(NSArray *aList, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aList];
    }];
}

- (void)getBlockListFromDB:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    
    NSArray * list = [EMClient.sharedClient.contactManager getBlackList];
    [self wrapperCallBack:result channelName:aChannelName error:nil object:list];
}

- (void)acceptInvitation:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    [EMClient.sharedClient.contactManager approveFriendRequestFromUser:username
                                                            completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
}

- (void)declineInvitation:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    [EMClient.sharedClient.contactManager declineFriendRequestFromUser:username
                                                            completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
}

- (void)getSelfIdsOnOtherPlatform:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    [EMClient.sharedClient.contactManager getSelfIdsOnOtherPlatformWithCompletion:^(NSArray *aList, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aList];
    }];
}


#pragma mark - EMContactManagerDelegate

- (void)friendshipDidAddByUser:(NSString *)aUsername {
    NSDictionary *map = @{
        @"type":@"onContactAdded",
        @"username":aUsername
    };
    [self.channel invokeMethod:EMMethodKeyOnContactChanged
                     arguments:map];
}

- (void)friendshipDidRemoveByUser:(NSString *)aUsername {
    NSDictionary *map = @{
        @"type":@"onContactDeleted",
        @"username":aUsername
    };
    [self.channel invokeMethod:EMMethodKeyOnContactChanged
                     arguments:map];
}

- (void)friendRequestDidReceiveFromUser:(NSString *)aUsername
                                message:(NSString *)aMessage {
    NSDictionary *map = @{
        @"type":@"onContactInvited",
        @"username":aUsername,
        @"reason":aMessage
    };
    [self.channel invokeMethod:EMMethodKeyOnContactChanged
                     arguments:map];
}

- (void)friendRequestDidApproveByUser:(NSString *)aUsername {
    NSDictionary *map = @{
        @"type":@"onFriendRequestAccepted",
        @"username":aUsername
    };
    [self.channel invokeMethod:EMMethodKeyOnContactChanged
                     arguments:map];
}

- (void)friendRequestDidDeclineByUser:(NSString *)aUsername {
    NSDictionary *map = @{
        @"type":@"onFriendRequestDeclined",
        @"username":aUsername
    };
    [self.channel invokeMethod:EMMethodKeyOnContactChanged
                     arguments:map];
}

@end
